nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/ConsumeMQTT.java

```diff
@@ -0,0 +1,341 @@
+            final MQTTQueueMessage mqttMessage = mqttQueue.peek();
+
+            Map<String, String> attrs = new HashMap<>();
+            attrs.put(BROKER_ATTRIBUTE_KEY, broker);
+            attrs.put(TOPIC_ATTRIBUTE_KEY, mqttMessage.getTopic());
+            attrs.put(QOS_ATTRIBUTE_KEY, String.valueOf(mqttMessage.getQos()));
+            attrs.put(IS_DUPLICATE_ATTRIBUTE_KEY, String.valueOf(mqttMessage.isDuplicate()));
+            attrs.put(IS_RETAINED_ATTRIBUTE_KEY, String.valueOf(mqttMessage.isRetained()));
+
+            messageFlowfile = session.putAllAttributes(messageFlowfile, attrs);
+
+            messageFlowfile = session.write(messageFlowfile, new OutputStreamCallback() {
+                @Override
+                public void process(final OutputStream out) throws IOException {
+                    out.write(mqttMessage.getPayload());
+                }
+            });
+
+            String transitUri = new StringBuilder(broker).append(mqttMessage.getTopic()).toString();
+            session.getProvenanceReporter().receive(messageFlowfile, transitUri);
+            session.transfer(messageFlowfile, REL_MESSAGE);
+            mqttQueue.remove(mqttMessage);
+            session.commit();
+        }
+    }
+
+    private class ConsumeMQTTCallback implements MqttCallback {
+
+        @Override
+        public void connectionLost(Throwable cause) {
+            logger.warn("Connection to " + broker + " lost", cause);
+            try {
+                reconnect();
+            } catch (MqttException e) {
+                logger.error("Connection to " + broker + " lost and callback re-connect failed.");
+            }
+        }
+
+        @Override
+        public void messageArrived(String topic, MqttMessage message) throws Exception {
+            logger.info("MQTT message arrived on topic:" + topic);
+            if (mqttQueue.size() >= maxQueueSize){
+                throw new IllegalStateException("The subscriber queue is full, cannot receive another message until the processor is scheduled to run.");
+            } else {
+                mqttQueue.add(new MQTTQueueMessage(topic, message));
+            }
+        }
+
+        @Override
+        public void deliveryComplete(IMqttDeliveryToken token) {
+            logger.warn("Received MQTT 'delivery complete' message to subscriber:"+ token);
+        }
+    }
+
+    // Public for testing
+    public void reconnect() throws MqttException {
+        synchronized (mqttClientConnectLock) {
+            if (!mqttClient.isConnected()) {
+                mqttClient = getMqttClient(broker, clientID, persistence);
+                mqttClient.setCallback(new ConsumeMQTTCallback());
+                mqttClient.connect(connOpts);
+                if(!subscribed){
+                    mqttClient.subscribe(topicFilter, qos);
+                    subscribed = true;
+                }
+            }
+        }
+    }
+
+    // Public for testing
+    public boolean isConnected(){
```

Comment:
"minimal in it's scope of impact" - may be for today, but you can not possibly answer this question for tomorrow. In fact by making it public one can argue you've created the 'scope' (however minimal it may be)  when it didn't even have to exist in the first place.
As for an example just get an instance of 'mqttClient' instance using the _Class.getDeclaredField_ dance and perform the same evaluation.
```
ConsumeMQTT mqtt = ...
Field f = mqtt.getClass().getDeclaredField("mqttClient");
f.setAccessible(true);
MqttClient mqttClient = f.get(mqtt);
// do your check on mqttClient
```
Put it in some private method within your test cases (some abstract class or some util class) and you're done.